WebGPURenderer: Add Offscreen Support

RenaudRohlinger opened the pull request:
This PR fix the use of the `WebGPURenderer` in a worker context.


A small issue is that the WebGPURenderer is intricately tied to direct imports from 'three' making it complex to create an example without changing the configuration of the project.
Web browsers do not currently support the importmap feature in worker contexts. Alternatives would be to update the rollup configuration or use relatives paths. (`Uncaught (in promise) TypeError: Failed to resolve module specifier "three". Relative references must start with either "/", "./", or "../".`)

I tested on my local with another project on both backends and it works fine.

_This contribution is funded by [Utsubo](https://utsubo.com/)_

RenaudRohlinger committed:
add offscreen support

sunag commented:
> Web browsers do not currently support the importmap feature in worker contexts. Alternatives would be to update the rollup configuration or use relatives paths. (Uncaught (in promise) TypeError: Failed to resolve module specifier "three". Relative references must start with either "/", "./", or "../".)

Sorry, I didn't quite understand the relationship between the description and the [commit](https://github.com/mrdoob/three.js/pull/27520/commits/5d9dee0c7439ee04c993b773b30358bfa91d7341).

RenaudRohlinger commented:
I wanted to add an example for `webgpu_offscreen` but it's currently not possible with the WebGPURenderer being in `examples/jsm`.

The PR by itself simply adds Worker support. 😁

sunag commented:
Maybe you didn't choose other commits by mistake?

![image](https://github.com/mrdoob/three.js/assets/502810/87dc6e71-3587-4cea-9eb0-9ded53d740d1)

RenaudRohlinger committed:
refactor statics webgpu

RenaudRohlinger committed:
navigator.gpu should be enough for initial check

RenaudRohlinger committed:
fix webgl context ktx2loader

RenaudRohlinger committed:
use isAvailable in webgpurenderer

RenaudRohlinger commented:
Indeed that was my bad @sunag. Basically before we were injecting a promise outside of the scope of the module which is something not recommend and which was breaking things such as web workers.

If that's ok with you I converted the value of `isAvailable` with `navigator.gpu !== undefined` as it's basically enough in 99% of the cases to know if WebGPU is available or not. Otherwise we would need to make `WebGPURenderer()` a promise which is something we don't want as it would be a pretty big breaking change.

Then the second issue was when `requestAdapter` was really needed in order to check the features available with `hasFeatures`. For that I converted `hasFeatures` to an async function and added a new method in `KTX2Loader` `detectSupportAsync`, so that the legacy Renderer is not affected by these changes. Both backends works fine within and outside the workers this way.

RenaudRohlinger committed:
cleanup and should fix pupeeter

RenaudRohlinger committed:
return promise

sunag commented on examples/jsm/renderers/webgpu/WebGPUBackend.js:
```diff
@@ -1071,9 +1062,9 @@ class WebGPUBackend extends Backend {
 	}
 
-	hasFeature( name ) {
+	async hasFeature( name ) {
```
I think we could create a new `hasFeatureAsync` to be used in `detectSupportAsync` and leave `hasFeature` as `chained method` returning `this`, we can use just the internal adapter `this.adapter` for `hasFeature`. We will probably have to have an `warn` the user saying if `this.adapter` has not been initialized and suggest that he can use the `detectSupportAsync` function instead.

RenaudRohlinger replied:
Thanks! Applied the feedbacks. 👍

sunag commented:
Thanks for the more detailed explanation. :)

RenaudRohlinger committed:
feedbacks

marcofugaro commented on examples/jsm/capabilities/WebGPU.js:
```diff
@@ -26,6 +23,12 @@ class WebGPU {
 
 	}
 
+	static async getStaticAdapter() {
+
+		return await isAvailable;
+
+	}
+
```
unnecessary async/await here (isAvailable is already awaited when it's set)

marcofugaro replied:
although my suggestion would be to keep the async/await here and remove the top level await instead

RenaudRohlinger replied:
Removed the async/await. To remove the top level await was my initial intention but it's actually necessary to have an accurate result from `navigator.gpu.requestAdapter()`. For instance Puppeeter will use the WebGPU backend as `isAvailable = navigator.gpu !== undefined` is true, while not being able to run WebGPU and so break all Github actions.

marcofugaro replied:
ok, maybe this is for another PR

marcofugaro commented on examples/jsm/capabilities/WebGPU.js:
```diff
@@ -26,6 +23,12 @@ class WebGPU {
 
 	}
```
the `WebGPU.isAvailable()` signature is changed, maybe do `return Boolean(isAvailable)` or create a new variable for the GPUAdapter

marcofugaro replied:
hmm, I think creating a new variable is cleaner

RenaudRohlinger committed:
remove unecessary async

RenaudRohlinger committed:
removed oversight

RenaudRohlinger commented on examples/webgpu_instance_mesh.html:
```diff
@@ -87,7 +87,7 @@
 				//
 
-				renderer = new WebGPURenderer( { antialias: true } );
+				renderer = await new WebGPURenderer( { antialias: true } );
```
Ah indeed! Thanks for that.

sunag committed:
cleanup
